### PR TITLE
Compile command

### PR DESCRIPTION
**Pull request by dav009**
# TL;DR
Adding a compile cmd to let users validate workflows without registration.

## Type
- [ ] Bug Fix
- [x] Feature
- [ ] Plugin

## Are all requirements met?

- [x] Code completed
- [ ] Smoke tested
- [x] Unit tests added
- [ ] Code documentation added
- [ ] Any pending items have an associated Issue

## Complete description

Usage: `flytectl compile --file  flyte-package.tgz`
Output sample:
```
Successfuly extracted package...
Processing Protobuff files...
Task: /tmp/register829973350/0_example.example.generate_normal_df_1.pb
Task: /tmp/register829973350/1_example.example.compute_stats_1.pb
Workflow: /tmp/register829973350/2_example.example.wf_2.pb
Launch Plan: /tmp/register829973350/3_example.example.wf_3.pb
Compiling tasks...
Compiling workflow: example.example.wf
All Workflows compiled successfully!

Summary:
1  workflows found in package
2  Tasks found in package
1  Launch plans found in package
```

Important changes:
- Imported flytepropeller's compiler (thus changes in go.mod)
- upgraded Go from 1.17 -> 1.18 (this is needed as flytepropeller seems to use a library that does not work with 1.17)
- Added a new command to cobra (`compile`)
- Added some minimal tests
 
## Tracking Issue
https://github.com/flyteorg/flyte/issues/2141

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by dav009** [comment]
@evalsocket  I can't run the the Lint job via https://github.com/nektos/act . 
Hopefully I fixed all the errors now, if you could re-run 🙏

**Comment by dav009** [comment]
@evalsocket  added some extra tests. However they do not address all codecov warnings. 
Any suggestions ? TA!

**Comment by evalsocket** [comment]
@dav009 codecov is pass, You can just fix the lint error

**Comment by dav009** [comment]
@evalsocket  re:re TA!

**Comment by dav009** [comment]
@evalsocket  sorry tiny nit. 
re re 🙏

File: cmd/compile/compile.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+	for _, pbFilePath := range fileList {
+		// deserializing packaged protocolbuffers
+		// based on filename as defined in flytekit
+		// https://github.com/flyteorg/flytekit/blob/master/flytekit/tools/serialize_helpers.py#L102
+		// filename suffix tell us how to deserialize file
+		// _1.pb : are Tasks protobufs
+		// _2.pb : are workflows protobufs
+		// _3.pb are Launchplans protobufs
+
```
Comment (pmahindrakar-oss) [inline cmd/compile/compile.go:58]: Instead of having this logic, can we use the similar logic as used during registration  https://github.com/flyteorg/flytectl/blob/5a17906593f9c6f4bcdb57268037aab768eb0724/cmd/register/register_util.go#L95

File: cmd/compile/compile.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+		Long:  compileLong,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			packageFilePath, _ := cmd.Flags().GetString("file")
+			return compileFromPackage(packageFilePath)
+		},
+	}
+	compile.Flags().StringVarP(&file, "file", "f", "", "path to file with packaged workflow..")
+	err := compile.MarkFlagRequired("file")
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (dav009) [reply cmd/compile/compile.go]: Hey, @pmahindrakar-oss  I had to change other functions that were only returning `cobra.Command` to return `cobra.Command, error` hope thats ok on your side.

**Comment by dav009** [comment]
@pmahindrakar-oss re:re 🙏   I think I addressed all of your comments .

File: cmd/compile/compile.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	return nil
+}
+
+const (
+	compileShort = `Validates your flyte packages without registration needed.`
+	compileLong  = `Validates your flyte packages without registration needed.  Validation is done by compiling your workflow's PB files. This ensures type safety and composition without needing a running flyte cluster.
+
+::
+
+ flytectl compile --file my-flyte-package.tgz
```
Comment (pmahindrakar-oss) [inline cmd/compile/compile.go:113]: Nit: check the other command on the example doc strings since these get published on the docs.flyte.org so the more detailed the better

**Review by pmahindrakar-oss — DISMISSED** [review]
Minor nit regarding docs.

File: cmd/compile/compile.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	compile := &cobra.Command{
+		Use:   "compile",
+		Short: compileShort,
+		Long:  compileLong,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			packageFilePath, _ := cmd.Flags().GetString("file")
+			return compileFromPackage(packageFilePath)
+		},
+	}
+	compile.Flags().StringVarP(&file, "file", "f", "", "path to a flyte package file. Flyte packages are tgz files generated by pyflyte or jflyte.")
```
Comment (pmahindrakar-oss) [inline cmd/compile/compile.go:130]: Also can we use a compile  config instead of just having one flag file: Check the other commands for example .

**Comment by dav009** [comment]
@pmahindrakar-oss :
- Changed docstrings trying to follow what I saw in the other cmds in the repo
- Added a config file for compile cmd. Also added the go-generated files associated with it
- Refactored tests to reflect this

**Review by pmahindrakar-oss — APPROVED** [review]

**Comment by pmahindrakar-oss** [comment]
Thanks very much for making those changes @dav009 .
